public/app/features/dashboard/components/SaveDashboard/useDashboardSave.tsx

```diff
@@ -8,12 +8,63 @@ import appEvents from 'app/core/app_events';
 import { updateLocation } from 'app/core/reducers/location';
 import { DashboardModel } from 'app/features/dashboard/state';
 import { saveDashboard as saveDashboardApiCall } from 'app/features/manage-dashboards/state/actions';
+import { getBackendSrv } from 'app/core/services/backend_srv';
+import config from 'app/core/config';
+
+const saveLibraryPanels = (saveModel: any, folderId: number) => {
+  // Check if there are any new library panels that need to be created first
+  const panelPromises = [];
+  for (const [i, panel] of saveModel.panels.entries()) {
+    if (!panel.libraryPanel) {
+      continue;
+    } else if (panel.libraryPanel && panel.libraryPanel.uid === undefined) {
+      panel.libraryPanel.name = panel.title;
+      panelPromises.push(
+        getBackendSrv()
+          .addLibraryPanel(panel, folderId!)
+          .then((returnedPanel) => {
+            saveModel.panels[i] = {
+              id: returnedPanel.model.id,
+              gridPos: returnedPanel.model.gridPos,
+              libraryPanel: {
+                uid: returnedPanel.uid,
+                name: returnedPanel.name,
+              },
+            };
+          })
+      );
+    } else {
+      // For now, update library panels. Implement "Update panel instances" modal later.
+      panelPromises.push(
+        getBackendSrv()
+          .updateLibraryPanel(panel, folderId!)
+          .then((returnedPanel) => {
+            saveModel.panels[i] = {
+              id: returnedPanel.model.id,
+              gridPos: returnedPanel.model.gridPos,
+              libraryPanel: {
+                uid: returnedPanel.uid,
+                name: returnedPanel.name,
+              },
+            };
+          })
+      );
+    }
+  }
+
+  return Promise.all(panelPromises).then(() => saveModel);
+};
 
 const saveDashboard = async (saveModel: any, options: SaveDashboardOptions, dashboard: DashboardModel) => {
   let folderId = options.folderId;
   if (folderId === undefined) {
-    folderId = dashboard.meta.folderId || saveModel.folderId;
+    folderId = dashboard.meta.folderId ?? saveModel.folderId;
   }
+
+  if (config.featureToggles.panelLibrary) {
+    await saveLibraryPanels(saveModel, folderId!);
+  }
```

Comment:
this is the wrong place, saving a library panel should be a specific action done from the panel edit mode. As is only for one panel at a time

Reply:
Agree. This was always just a temporary measure

Reply:
Just to clarify something: So the only way to save a library panel should be from the panel edit view?
If that's the case, if a user creates a new library panel - which automatically gets added to the current dashboard - and the user saves the dashboard before that panel is explicitly saved, should a modal be presented warning the user that they have unsaved library panels or something to that effect?